From d5eabd5ff34baffa8fc2e72f1d3715c1eeb83360 Mon Sep 17 00:00:00 2001
From: Brian Scholer <1260690+briantist@users.noreply.github.com>
Date: Mon, 3 Jul 2023 13:16:30 -0400
Subject: [PATCH] expand Vault CI matrix, announce deprecation of Vault dynamic
 SSH keys (#1023)

* expand Vault matrix

* turn off fail-fast

* add color to pytest output

* Fix azure result for newer Vault versions

Co-authored-by: Tyler Hess <tylerlhess@gmail.com>

* try to fix k8s integration failure in 1.9.x

* fix entity integration error

* update azure result version compare

* skip removed dynamic ssh key functionality in Vault >=1.13

* gimme a reason

* blackify

* decorator order

* try going back to enterprise versions

* revert enterprise changes

* include test type in coverage filename

* update codecov config, post comment with no changes

---------

Co-authored-by: Tyler Hess <tylerlhess@gmail.com>
---
 .github/workflows/lint-and-test.yml           | 29 ++++++++++++-------
 codecov.yml                                   |  7 ++---
 hvac/api/secrets_engines/ssh.py               |  6 ++--
 .../api/auth_methods/test_kubernetes.py       |  4 ++-
 .../api/secrets_engines/test_azure.py         |  3 +-
 .../api/secrets_engines/test_identity.py      |  5 +++-
 .../api/secrets_engines/test_ssh.py           | 11 ++++++-
 7 files changed, 45 insertions(+), 20 deletions(-)

diff --git a/hvac/api/secrets_engines/ssh.py b/hvac/api/secrets_engines/ssh.py
index 3e81f3a..10826d4 100644
--- a/hvac/api/secrets_engines/ssh.py
+++ b/hvac/api/secrets_engines/ssh.py
@@ -14,13 +14,14 @@ class Ssh(VaultApiBase):
     Reference: https://www.vaultproject.io/api-docs/secret/ssh
     """
 
+    # TODO: deprecate all dynamic SSH keys methods from hvac
     def create_or_update_key(
         self,
         name="",
         key="",
         mount_point=DEFAULT_MOUNT_POINT,
     ):
-        """This endpoint updates a named key.
+        """This endpoint updates a named key. This method uses deprecated functionality that was removed in Vault 1.13.0.
 
         :param name: Specifies the name of the key to create.
         :type name: str | unicode
@@ -46,12 +47,13 @@ class Ssh(VaultApiBase):
             json=params,
         )
 
+    # TODO: deprecate all dynamic SSH keys methods from hvac
     def delete_key(
         self,
         name="",
         mount_point=DEFAULT_MOUNT_POINT,
     ):
-        """This endpoint deletes a named key.
+        """This endpoint deletes a named key. This method uses deprecated functionality that was removed in Vault 1.13.0.
 
         :param name: Specifies the name of the key to delete.
         :type name: str | unicode
diff --git a/tests/integration_tests/api/secrets_engines/test_azure.py b/tests/integration_tests/api/secrets_engines/test_azure.py
index 862f20b..7ffa36b 100644
--- a/tests/integration_tests/api/secrets_engines/test_azure.py
+++ b/tests/integration_tests/api/secrets_engines/test_azure.py
@@ -111,7 +111,8 @@ class TestAzure(HvacIntegrationTestCase, TestCase):
         }
         if utils.vault_version_ge("1.9.0"):
             read_expected_response["root_password_ttl"] = 0
-            read_expected_response["use_microsoft_graph_api"] = False
+            if utils.vault_version_lt("1.12.0"):
+                read_expected_response["use_microsoft_graph_api"] = False
         self.assertEqual(
             first=read_expected_response,
             second=read_configuration_response,
diff --git a/tests/integration_tests/api/secrets_engines/test_ssh.py b/tests/integration_tests/api/secrets_engines/test_ssh.py
index b65f42e..b7991a8 100644
--- a/tests/integration_tests/api/secrets_engines/test_ssh.py
+++ b/tests/integration_tests/api/secrets_engines/test_ssh.py
@@ -1,9 +1,10 @@
 import logging
-from unittest import TestCase
+from unittest import TestCase, skipIf
 
 from parameterized import parameterized, param
 
 from tests.utils.hvac_integration_test_case import HvacIntegrationTestCase
+from tests import utils
 
 
 class TestPki(HvacIntegrationTestCase, TestCase):
@@ -22,6 +23,7 @@ class TestPki(HvacIntegrationTestCase, TestCase):
         self.client.sys.disable_secrets_engine(path=self.TEST_MOUNT_POINT)
         super().tearDown()
 
+    # TODO: deprecate all dynamic SSH keys methods from hvac
     @parameterized.expand(
         [
             param(
@@ -29,6 +31,9 @@ class TestPki(HvacIntegrationTestCase, TestCase):
             ),
         ]
     )
+    @skipIf(
+        utils.vault_version_ge("1.13.0"), reason="Vault 1.13.0 dropped this feature."
+    )
     def test_create_key(self, label, raises=False, exception_message=""):
         with open(self.PRIVATE_SSH_KEY) as key_file:
             private_key = key_file.read()
@@ -44,6 +49,7 @@ class TestPki(HvacIntegrationTestCase, TestCase):
             second=204,
         )
 
+    # TODO: deprecate all dynamic SSH keys methods from hvac
     @parameterized.expand(
         [
             param(
@@ -51,6 +57,9 @@ class TestPki(HvacIntegrationTestCase, TestCase):
             ),
         ]
     )
+    @skipIf(
+        utils.vault_version_ge("1.13.0"), reason="Vault 1.13.0 dropped this feature."
+    )
     def test_delete_key(self, label, raises=False, exception_message=""):
         with open(self.PRIVATE_SSH_KEY) as key_file:
             private_key = key_file.read()
-- 
2.42.0

